xapian-core/cluster/cluster.cc

```diff
@@ -604,3 +616,44 @@ Cluster::Internal::recalculate()
     }
     centroid.divide(size());
 }
+
+StemStopper::StemStopper(const Stem &stemmer_)
+    : stem_action(STEM_SOME), stemmer(stemmer_)
+{
+    LOGCALL_CTOR(API, "StemStopper", stemmer_);
+}
+
+void
+StemStopper::set_stemming_strategy(stem_strategy strategy)
+{
+    LOGCALL_VOID(API, "StemStopper::set_stemming_strategy", strategy);
+    stem_action = strategy;
+}
+
+string
+StemStopper::get_description() const
+{
+    string desc("Xapian::StemStopper(");
+    unordered_set<string>::const_iterator i;
+    for (i = stop_words.begin(); i != stop_words.end(); ++i) {
+	if (i != stop_words.begin()) desc += ' ';
+	desc += *i;
+    }
+    desc += ')';
+    return desc;
+}
+
+void
+StemStopper::add(const string &term)
+{
+    LOGCALL_VOID(API, "StemStopper::add", term);
+    const string &stemmed_term = 'Z' + stemmer(term);
+    if (stem_action == STEM_NONE)
+	stop_words.insert(term);
+    else if (stem_action == STEM_ALL_Z)
+	stop_words.insert(stemmed_term);
+    else {
+	stop_words.insert(term);
+	stop_words.insert(stemmed_term);
+    }
+}
```

Comment:
This handles `STEM_ALL` the same as `STEM_SOME` which is incorrect.

I'd suggest handling an `enum` with a `switch` rather than an `if` cascade.  That way we'll get a warning (at least with GCC) if a new `enum` value is added and nobody adds the code here to handle it.

This also always calls the stemmer, but under `STEM_NONE` the result is never used, so that's wasted effort.

And finally, it's probably better stylistically to use curly brackets on all or none of the actions in an `if`/`else if`/`else` chain.